software/spmd/bsg_cuda_lite_runtime/striped/kernel_striped.cpp

```diff
@@ -0,0 +1,36 @@
+//This is an empty kernel
+
+#include "bsg_manycore.h"
+#include "bsg_set_tile_x_y.h"
+#include "bsg_striped_array.hpp"
+
+#define BSG_TILE_GROUP_X_DIM bsg_tiles_X
+#define BSG_TILE_GROUP_Y_DIM bsg_tiles_Y
+#include "bsg_tile_group_barrier.h"
+INIT_TILE_GROUP_BARRIER(r_barrier, c_barrier, 0, bsg_tiles_X-1, 0, bsg_tiles_Y-1);
+
+using namespace bsg_manycore;
+
+struct foo {
+    int bar;
+    int baz;
+};
+
+using Array = TileGroupStripedArray<struct foo, 16, bsg_tiles_X, bsg_tiles_Y, 2>;
+
+extern "C" int  __attribute__ ((noinline)) kernel_striped() {
+    Array data;
+    if (bsg_id == 0) {
+        for (int i = 0; i < data.size(); i++) {
+            data[i].bar = 2*i;
+            data[i].baz = 2*i+1;
+        }
+    }
+
+    bsg_tile_group_barrier(&r_barrier, &c_barrier);
```

Comment:
This test correctly tests when tile with ID 0 writes to other tiles' memories. Due to the lack of `volatile` above, I believe this will fail down the line if you synchronize, have some other tiles write to their local addresses, and then you have tile 0 read again. It's likely that the value will be cached in a register and you won't see the updated value (I ran into this problem a while ago so that's how I know 😝).

Reply:
That's the point of a memory fence, as we were discussing on Thursday, right? After a fence, it'll go read the remote cores again. 

I think I'm with Max's comment - it's better to have the programmer specify the volatility, unless there's some fundamental reason.

Reply:
The compiler memory fence that we discussed on Thursday is to prevent the compiler from reordering reads/writes between sync points. I don't think it will necessarily make compiler decide to reread the value from remote tiles. 
I'd say the fundamental reason you want volatile by default is for correctness. It's a massive gotcha where you have to force a remote load. Your results will differ a lot between debug builds and release builds. 
I think that we need to make this shared array be easy to program with correctly. Someone who wants to optimize the use of the array should manually copy the values out into a local variable and then perform whatever they need to do. Otherwise this StripedArray class doesn't really conceptually represent a striped array - it represents some weird mess of having several independent copies of the same element (except it's undefined how many copies), which sometimes interact (except its' undefined when they interact), and which depend entirely on your compiler's register allocator.

Reply:
> Otherwise this StripedArray class doesn't really conceptually represent a striped array - it represents some weird mess of having several independent copies of the same element (except it's undefined how many copies), which sometimes interact (except its' undefined when they interact), and which depend entirely on your compiler's register allocator.

I'm not sure I agree with this characterization. It doesn't make several independent copies of an array; it builds an array whose elements are interleaved across a group of tiles (a group of tiles is a "tile group," which is in the name of the class, and is a well defined thing in the context of our software stack). The template parameters allow the programmer to control how the elements are interleaved. 

Other than that, it has exactly the same semantics as a normal array and to be honest I'm pretty sure that's what we want by default. @save-buffer you're correct that it will often be necessary to mark this data as volatile, but suppose you just want more memory for an array than the 4KB on a single tile? Maybe there's not enough parallel work to justify actually using all the compute. In such a situation, you probably would not want the array to be treated as volatile all the time. 

> I think that we need to make this shared array be easy to program with correctly. Someone who wants to optimize the use of the array should manually copy the values out into a local variable and then perform whatever they need to do.

The thing is, it's way easier for a programmer to discover that they're getting the wrong answer than it is to find out that their performance is being degraded by some helper class that's marking everything as volatile by default. 

Further, if some poor programmer who knows what they're doing really does want to remove volatile, they'll have to copy and paste the original class and remove all the volatile keywords (or paste a bunch of calls to remove_volatile in their kernel). 

All that being said, catering to the common use case is a reasonable thing to do. I think the best way to achieve that is to make a wrapper that reuses this class but also adds volatile to the returned references. That way we get the best of both worlds.